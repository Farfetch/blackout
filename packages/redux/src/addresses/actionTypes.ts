/**
 * Action type dispatched when the fetch prediction request fails.
 */
export const FETCH_ADDRESS_PREDICTIONS_FAILURE =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_FAILURE';
/**
 * Action type dispatched when the fetch prediction request starts.
 */
export const FETCH_ADDRESS_PREDICTIONS_REQUEST =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_REQUEST';
/**
 * Action type dispatched when the fetch prediction request succeeds.
 */
export const FETCH_ADDRESS_PREDICTIONS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_SUCCESS';

/**
 * Action type dispatched when the fetch prediction details request fails.
 */
export const FETCH_ADDRESS_PREDICTION_FAILURE =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_FAILURE';
/**
 * Action type dispatched when the fetch prediction details request starts.
 */
export const FETCH_ADDRESS_PREDICTION_REQUEST =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_REQUEST';
/**
 * Action type dispatched when the fetch prediction details request succeeds.
 */
export const FETCH_ADDRESS_PREDICTION_SUCCESS =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_SUCCESS';

/**
 * Action type dispatched when resetting the address prediction.
 */
export const RESET_ADDRESS_PREDICTIONS =
  '@farfetch/blackout-redux/RESET_ADDRESS_PREDICTIONS';
