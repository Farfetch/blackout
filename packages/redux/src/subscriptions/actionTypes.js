/**
 * @module subscriptions/actionTypes
 * @category Subscriptions
 * @subcategory Actions
 */

/** Action type dispatched when the fetch subscription packages request fails. */
export const FETCH_SUBSCRIPTION_PACKAGES_FAILURE =
  '@farfetch/blackout-redux/FETCH_SUBSCRIPTION_PACKAGES_FAILURE';
/** Action type dispatched when the fetch subscription packages request starts. */
export const FETCH_SUBSCRIPTION_PACKAGES_REQUEST =
  '@farfetch/blackout-redux/FETCH_SUBSCRIPTION_PACKAGES_REQUEST';
/** Action type dispatched when the fetch subscription packages request succeeds. */
export const FETCH_SUBSCRIPTION_PACKAGES_SUCCESS =
  '@farfetch/blackout-redux/FETCH_SUBSCRIPTION_PACKAGES_SUCCESS';

/** Action type dispatched when the fetch user subscriptions request fails. */
export const FETCH_USER_SUBSCRIPTIONS_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_SUBSCRIPTIONS_FAILURE';
/** Action type dispatched when the fetch user subscriptions request starts. */
export const FETCH_USER_SUBSCRIPTIONS_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_SUBSCRIPTIONS_REQUEST';
/** Action type dispatched when the fetch user subscriptions request succeeds. */
export const FETCH_USER_SUBSCRIPTIONS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_SUBSCRIPTIONS_SUCCESS';

/** Action type dispatched when the put user subscriptions request fails. */
export const UPDATE_USER_SUBSCRIPTIONS_FAILURE =
  '@farfetch/blackout-redux/UPDATE_USER_SUBSCRIPTIONS_FAILURE';
/** Action type dispatched when the put user subscriptions request starts. */
export const UPDATE_USER_SUBSCRIPTIONS_REQUEST =
  '@farfetch/blackout-redux/UPDATE_USER_SUBSCRIPTIONS_REQUEST';
/** Action type dispatched when the put user subscriptions request succeeds. */
export const UPDATE_USER_SUBSCRIPTIONS_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_USER_SUBSCRIPTIONS_SUCCESS';

/** Action type dispatched by [reset]{@link module:subscriptions/actions.reset} thunk. */
export const RESET_SUBSCRIPTIONS =
  '@farfetch/blackout-redux/RESET_SUBSCRIPTIONS';

/** Action type dispatched when the unsubscribe all subscriptions request fails. */
export const UNSUBSCRIBE_ALL_SUBSCRIPTIONS_FAILURE =
  '@farfetch/blackout-redux/UNSUBSCRIBE_ALL_SUBSCRIPTIONS_FAILURE';
/** Action type dispatched when the unsubscribe all subscriptions request starts. */
export const UNSUBSCRIBE_ALL_SUBSCRIPTIONS_REQUEST =
  '@farfetch/blackout-redux/UNSUBSCRIBE_ALL_SUBSCRIPTIONS_REQUEST';
/** Action type dispatched when the unsubscribe all subscriptions request succeeds. */
export const UNSUBSCRIBE_ALL_SUBSCRIPTIONS_SUCCESS =
  '@farfetch/blackout-redux/UNSUBSCRIBE_ALL_SUBSCRIPTIONS_SUCCESS';

/** Action type dispatched when the unsubscribe recipient from topic request fails. */
export const UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_FAILURE =
  '@farfetch/blackout-redux/UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_FAILURE';
/** Action type dispatched when the unsubscribe recipient from topic request starts. */
export const UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUEST =
  '@farfetch/blackout-redux/UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUEST';
/** Action type dispatched when the unsubscribe recipient from topic request succeeds. */
export const UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_SUCCESS =
  '@farfetch/blackout-redux/UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_SUCCESS';

/** Action type dispatched when the application wants to clear a specific unsubscribe recipient from topic request state from the store. */
export const CLEAR_UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUEST =
  '@farfetch/blackout-redux/CLEAR_UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUEST';

/** Action type dispatched when the application wants to clear all unsubscribe recipient from topic requests state from the store. */
export const CLEAR_ALL_UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUESTS =
  '@farfetch/blackout-redux/CLEAR_ALL_UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUESTS';
