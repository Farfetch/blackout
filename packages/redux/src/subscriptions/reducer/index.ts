import { combineReducers } from 'redux';
import subscriptionPackagesReducer, {
  subscriptionPackagesEntitiesMapper,
} from './subscriptionPackages.js';
import userSubscriptionReducer from './userSubscriptions.js';

export const entitiesMapper = {
  ...subscriptionPackagesEntitiesMapper,
};

const reducers = combineReducers({
  user: userSubscriptionReducer,
  packages: subscriptionPackagesReducer,
});

export default reducers;
