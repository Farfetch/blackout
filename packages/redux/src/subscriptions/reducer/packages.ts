import * as actionTypes from '../actionTypes';
import { AnyAction, combineReducers } from 'redux';
import type { PackagesState } from '../types';
import type { ReducerSwitch } from '../../types';

export const INITIAL_STATE: PackagesState = {
  error: null,
  isLoading: false,
  result: null,
};

const error = (state = INITIAL_STATE.error, action: AnyAction) => {
  switch (action.type) {
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action: AnyAction) => {
  switch (action.type) {
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_REQUEST:
      return true;
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_SUCCESS:
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_FAILURE:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const result = (state = INITIAL_STATE.result, action: AnyAction) => {
  switch (action.type) {
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_SUCCESS:
      return action.payload.result;
    default:
      return state;
  }
};

export const getPackagesError = (
  state: PackagesState | undefined,
): PackagesState['error'] | undefined => state?.error;
export const getPackagesIsLoading = (
  state: PackagesState | undefined,
): PackagesState['isLoading'] | undefined => state?.isLoading;
export const getPackages = (
  state: PackagesState | undefined,
): PackagesState['result'] | undefined => state?.result;

const reducers = combineReducers({
  error,
  isLoading,
  result,
});

/**
 * Reducer for Subscription Packages state.
 *
 * @param state - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */

const packagesSubscriptionReducer: ReducerSwitch<PackagesState, AnyAction> = (
  state = INITIAL_STATE,
  action,
): PackagesState => {
  return reducers(state, action);
};

export default packagesSubscriptionReducer;
