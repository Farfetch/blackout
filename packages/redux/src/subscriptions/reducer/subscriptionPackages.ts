import * as actionTypes from '../actionTypes.js';
import type { AnyAction } from 'redux';
import type { StoreState } from '../../types/index.js';
import type { SubscriptionsState } from '../types/index.js';

export const INITIAL_STATE: Record<
  string,
  SubscriptionsState['packages'][string]
> = {};

const subscriptionPackagesReducer = (
  state = INITIAL_STATE,
  action: AnyAction,
) => {
  switch (action.type) {
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_REQUEST:
      return {
        ...state,
        [action.meta.hash]: {
          result: null,
          isLoading: true,
          error: null,
        },
      };
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_SUCCESS:
      return {
        ...state,
        [action.meta.hash]: {
          result: action.payload.result,
          isLoading: false,
          error: null,
        },
      };
    case actionTypes.FETCH_SUBSCRIPTION_PACKAGES_FAILURE:
      return {
        ...state,
        [action.meta.hash]: {
          result: null,
          isLoading: false,
          error: action.payload.error,
        },
      };
    case actionTypes.RESET_SUBSCRIPTION_PACKAGES:
      return INITIAL_STATE;
    default:
      return state;
  }
};

const removeSubscriptionPackagesFromEntities = (
  state: NonNullable<StoreState['entities']>,
) => ({ ...state, subscriptionPackages: undefined });

export const subscriptionPackagesEntitiesMapper = {
  [actionTypes.RESET_SUBSCRIPTION_PACKAGES]:
    removeSubscriptionPackagesFromEntities,
};

export default subscriptionPackagesReducer;
