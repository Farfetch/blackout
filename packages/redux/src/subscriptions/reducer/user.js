import * as actionTypes from '../actionTypes';
import { combineReducers } from 'redux';

export const INITIAL_STATE = {
  error: null,
  isLoading: false,
  result: null,
  unsubscribeRecipientFromTopicRequests: {},
};

const error = (
  state = INITIAL_STATE.error,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.UNSUBSCRIBE_FROM_SUBSCRIPTION_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UNSUBSCRIBE_FROM_SUBSCRIPTION_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UNSUBSCRIBE_FROM_SUBSCRIPTION_REQUEST:
      return true;
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_SUCCESS:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_SUCCESS:
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.UNSUBSCRIBE_FROM_SUBSCRIPTION_SUCCESS:
    case actionTypes.UNSUBSCRIBE_FROM_SUBSCRIPTION_FAILURE:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const result = (
  state = INITIAL_STATE.result,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_SUCCESS:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_SUCCESS:
      return action.payload;
    case actionTypes.UNSUBSCRIBE_FROM_SUBSCRIPTION_SUCCESS:
      return INITIAL_STATE.result;
    case actionTypes.UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_SUCCESS: {
      const { subscriptionId, topicId, recipientId } = action.payload;

      const subscriptionIndex = state?.findIndex(
        subscription => subscription.id === subscriptionId,
      );

      if (subscriptionIndex > -1) {
        const subscription = state[subscriptionIndex];

        const topicIndex = subscription.topics?.findIndex(
          topic => topic.id === topicId,
        );

        if (topicIndex > -1) {
          const topic = subscription.topics[topicIndex];

          const recipientIndex = topic.channels?.findIndex(
            channel => channel.id === recipientId,
          );

          if (recipientIndex > -1) {
            // Found recipient in subscription.
            // Create a new state that does not include this recipientId

            const newTopicChannels = [...topic.channels];
            newTopicChannels.splice(recipientIndex, 1);

            const newTopics = [...subscription.topics];

            // If there are no channels, remove the topic altogether
            if (!newTopicChannels.length) {
              newTopics.splice(topicIndex, 1);
            } else {
              newTopics[topicIndex] = {
                ...newTopics[topicIndex],
                channels: newTopicChannels,
              };
            }

            const newState = [...state];

            // If there are no topics, remove the subscription altogether
            if (!newTopics.length) {
              newState.splice(subscriptionIndex, 1);
            } else {
              newState[subscriptionIndex] = {
                ...newState[subscriptionIndex],
                topics: newTopics,
              };
            }

            return newState;
          }
        }
      }
      return state;
    }
    default:
      return state;
  }
};

const unsubscribeRecipientFromTopicRequests = (
  state = INITIAL_STATE.unsubscribeRecipientFromTopicRequests,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUEST: {
      const trackRequestState = action.meta?.trackRequestState;

      if (trackRequestState) {
        const { subscriptionId, topicId, recipientId } = action.payload;

        return {
          ...state,
          [recipientId]: {
            subscriptionId,
            topicId,
            isFetching: true,
            error: undefined,
            success: undefined,
          },
        };
      }

      return state;
    }

    case actionTypes.UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_SUCCESS: {
      const recipientId = action.payload?.recipientId;
      const requestState = state[recipientId];

      if (requestState) {
        return {
          ...state,
          [recipientId]: {
            ...requestState,
            isFetching: false,
            success: true,
          },
        };
      }

      return state;
    }

    case actionTypes.UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_FAILURE: {
      const error = action.payload?.error;
      const recipientId = action.payload?.recipientId;
      const requestState = state[recipientId];

      if (requestState) {
        return {
          ...state,
          [recipientId]: {
            ...requestState,
            isFetching: false,
            success: false,
            error,
          },
        };
      }

      return state;
    }

    case actionTypes.CLEAR_UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUEST: {
      const { recipientId } = action.payload;

      if (recipientId in state) {
        const newState = {
          ...state,
        };

        delete newState[recipientId];

        return newState;
      }

      return state;
    }

    case actionTypes.CLEAR_ALL_UNSUBSCRIBE_RECIPIENT_FROM_TOPIC_REQUESTS: {
      return INITIAL_STATE.unsubscribeRecipientFromTopicRequests;
    }

    default:
      return state;
  }
};

export const getSubscriptionsError = state => state.error;
export const getSubscriptionsIsLoading = state => state.isLoading;
export const getSubscriptions = state => state.result;
export const getUnsubscribeRecipientFromTopicRequests = state =>
  state.unsubscribeRecipientFromTopicRequests;

export default combineReducers({
  error,
  isLoading,
  result,
  unsubscribeRecipientFromTopicRequests,
});
