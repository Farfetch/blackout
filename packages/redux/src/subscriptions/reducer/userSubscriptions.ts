import * as actionTypes from '../actionTypes';
import * as authenticationActionTypes from '../../users/authentication/actionTypes';
import { type AnyAction, combineReducers, type Reducer } from 'redux';
import type {
  Subscription,
  SubscriptionTopic,
} from '@farfetch/blackout-client';
import type { SubscriptionsState } from '../types';

type UserSubscriptionsState = SubscriptionsState['user'];

export const INITIAL_STATE: UserSubscriptionsState = {
  error: null,
  isLoading: false,
  result: null,
  unsubscribeRecipientFromTopicRequests: {},
  updateSubscriptionsError: null,
};

const error = (state = INITIAL_STATE.error, action: AnyAction) => {
  switch (action.type) {
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_FAILURE:
      return action.payload.error;
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const updateSubscriptionsError = (
  state = INITIAL_STATE.updateSubscriptionsError,
  action: AnyAction,
) => {
  switch (action.type) {
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_FAILURE:
      return action.payload.error;
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_REQUEST:
      return INITIAL_STATE.updateSubscriptionsError;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action: AnyAction) => {
  switch (action.type) {
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_REQUEST:
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_REQUEST:
      return true;
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_SUCCESS:
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_SUCCESS:
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_FAILURE:
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_SUCCESS:
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_FAILURE:
      return false;
    default:
      return state;
  }
};

const result = (state = INITIAL_STATE.result, action: AnyAction) => {
  switch (action.type) {
    case actionTypes.FETCH_USER_SUBSCRIPTIONS_SUCCESS:
      return action.payload;
    case actionTypes.UPDATE_USER_SUBSCRIPTIONS_SUCCESS:
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_SUCCESS:
      return INITIAL_STATE.result;
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_TOPIC_RECIPIENT_SUCCESS: {
      const { subscriptionId, topicId, recipientId } = action.payload;

      if (!state) {
        return state;
      }

      const subscription = state.find(
        subscription => subscription.id === subscriptionId,
      );

      if (subscription) {
        const topic = subscription.topics.find(topic => topic.id === topicId);

        if (topic) {
          const recipient = topic.channels?.find(
            channel => channel.id === recipientId,
          );

          if (recipient) {
            // New topic channels without old recipient id
            const newTopicChannels = [
              ...topic.channels.filter(channel => channel.id !== recipientId),
            ];

            // update topic list
            const newTopics: SubscriptionTopic[] = [];

            if (!newTopicChannels.length) {
              // if result channel list are empty, then remove topic entry
              newTopics.push(
                ...subscription.topics.filter(topic => topic.id !== topicId),
              );
            } else {
              // replace old topic list replacing old topic list with new one
              newTopics.push(
                ...subscription.topics.map(topic => {
                  return {
                    ...topic,
                    channels:
                      topic.id === topicId ? newTopicChannels : topic.channels,
                  };
                }),
              );
            }

            const newState: Subscription[] = [];

            if (!newTopics.length) {
              // if result channel list are empty, then remove topic entry
              newState.push(
                ...state.filter(
                  subscription => subscription.id !== subscriptionId,
                ),
              );
            } else {
              // replace old topic list replacing old topic list with new one
              newState.push(
                ...state.map(subscription => {
                  return {
                    ...subscription,
                    topics:
                      subscription.id === subscriptionId
                        ? newTopics
                        : subscription.topics,
                  };
                }),
              );
            }

            return newState;
          }
        }
      }

      return state;
    }
    default:
      return state;
  }
};

const unsubscribeRecipientFromTopicRequests = (
  state = INITIAL_STATE.unsubscribeRecipientFromTopicRequests,
  action: AnyAction,
) => {
  switch (action.type) {
    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_TOPIC_RECIPIENT_REQUEST: {
      const trackRequestState = action.meta?.trackRequestState;

      if (trackRequestState) {
        const { subscriptionId, topicId, recipientId } = action.payload;

        return {
          ...state,
          [recipientId]: {
            subscriptionId,
            topicId,
            isFetching: true,
            error: undefined,
            success: undefined,
          },
        };
      }

      return state;
    }

    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_TOPIC_RECIPIENT_SUCCESS: {
      const recipientId = action.payload?.recipientId;
      const requestState = state[recipientId];

      if (requestState) {
        return {
          ...state,
          [recipientId]: {
            ...requestState,
            isFetching: false,
            success: true,
          },
        };
      }

      return state;
    }

    case actionTypes.UNSUBSCRIBE_SUBSCRIPTION_TOPIC_RECIPIENT_FAILURE: {
      const error = action.payload?.error;
      const recipientId = action.payload?.recipientId;
      const requestState = state[recipientId];

      if (requestState) {
        return {
          ...state,
          [recipientId]: {
            ...requestState,
            isFetching: false,
            success: false,
            error,
          },
        };
      }

      return state;
    }

    case actionTypes.CLEAR_UNSUBSCRIBE_SUBSCRIPTION_TOPIC_RECIPIENT_REQUEST: {
      const { recipientId } = action.payload;

      if (recipientId in state) {
        const newState = {
          ...state,
        };

        delete newState[recipientId];

        return newState;
      }

      return state;
    }

    case actionTypes.CLEAR_ALL_UNSUBSCRIBE_SUBSCRIPTION_TOPIC_RECIPIENT_FROM_REQUESTS: {
      return INITIAL_STATE.unsubscribeRecipientFromTopicRequests;
    }

    default:
      return state;
  }
};

export const getUserSubscriptionsError = (
  state: UserSubscriptionsState = INITIAL_STATE,
) => state.error;
export const getUserSubscriptionsIsLoading = (
  state: UserSubscriptionsState = INITIAL_STATE,
) => state.isLoading;
export const getUserSubscriptions = (
  state: UserSubscriptionsState = INITIAL_STATE,
) => state.result;
export const getUnsubscribeRecipientFromTopicRequests = (
  state: UserSubscriptionsState = INITIAL_STATE,
) => state.unsubscribeRecipientFromTopicRequests;
export const getUpdateSubscriptionsError = (
  state: UserSubscriptionsState = INITIAL_STATE,
) => state.updateSubscriptionsError;

const reducer = combineReducers({
  error,
  isLoading,
  result,
  unsubscribeRecipientFromTopicRequests,
  updateSubscriptionsError,
});

/**
 * Reducer for user subscription state.
 *
 * @param state  - Current redux state.
 * @param action - Action dispatched.
 *
 * @returns New state.
 */

const userSubscriptionsReducer: Reducer<UserSubscriptionsState> = (
  state = INITIAL_STATE,
  action,
) => {
  switch (action.type) {
    case actionTypes.RESET_USER_SUBSCRIPTIONS:
    case authenticationActionTypes.LOGOUT_SUCCESS:
    case authenticationActionTypes.LOGIN_SUCCESS:
    case authenticationActionTypes.REGISTER_SUCCESS:
    case authenticationActionTypes.FETCH_USER_SUCCESS:
      return INITIAL_STATE;

    default:
      return reducer(state, action);
  }
};

export default userSubscriptionsReducer;
