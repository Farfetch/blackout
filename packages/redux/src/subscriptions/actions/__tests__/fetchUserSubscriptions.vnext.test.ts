import * as actionTypes from '../../actionTypes.js';
import { fetchUserSubscriptionsVNext } from '../index.js';
import { find } from 'lodash-es';
import { getSubscriptionsVNext } from '@farfetch/blackout-client';
import { mockGetSubscriptionsVNext } from 'tests/__fixtures__/subscriptions/index.mjs';
import { mockStore } from '../../../../tests/index.js';
import reducer from '../../reducer/index.js';

jest.mock('@farfetch/blackout-client', () => {
  return {
    ...jest.requireActual('@farfetch/blackout-client'),
    getSubscriptionsVNext: jest.fn(),
  };
});

const randomAction = { type: 'this_is_a_random_action' };
const subscriptionsMockStore = (state = {}) =>
  mockStore({ getSubscriptionsVNext: reducer(undefined, randomAction) }, state);

describe('Subscriptions redux actions', () => {
  let store: ReturnType<typeof subscriptionsMockStore>;

  beforeEach(jest.clearAllMocks);

  describe('fetchUserSubscriptionsVNext() action creator', () => {
    beforeEach(() => {
      store = subscriptionsMockStore();
    });

    it('Should create the correct actions for when the get subscription fails', async () => {
      const expectedError = new Error('get subscriptions error');

      (getSubscriptionsVNext as jest.Mock).mockRejectedValueOnce(expectedError);

      await expect(
        async () =>
          await fetchUserSubscriptionsVNext(mockGetSubscriptionsVNext.query)(
            store.dispatch,
          ),
      ).rejects.toThrow(expectedError);

      expect(getSubscriptionsVNext).toHaveBeenCalledTimes(1);
      expect(getSubscriptionsVNext).toHaveBeenCalledWith(
        mockGetSubscriptionsVNext.query,
        undefined,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          { type: actionTypes.FETCH_USER_SUBSCRIPTIONS_REQUEST_VNEXT },
          {
            type: actionTypes.FETCH_USER_SUBSCRIPTIONS_FAILURE,
            payload: { error: expectedError },
          },
        ]),
      );
    });

    it('Should create the correct actions for when the get subscriptions is successful', async () => {
      (getSubscriptionsVNext as jest.Mock).mockResolvedValueOnce(
        mockGetSubscriptionsVNext.response,
      );

      await fetchUserSubscriptionsVNext(mockGetSubscriptionsVNext.query)(
        store.dispatch,
      ).then(clientResult => {
        expect(clientResult).toBe(mockGetSubscriptionsVNext.response);
      });

      const actionResults = store.getActions();

      expect(getSubscriptionsVNext).toHaveBeenCalledTimes(1);
      expect(getSubscriptionsVNext).toHaveBeenCalledWith(
        mockGetSubscriptionsVNext.query,
        undefined,
      );
      expect(actionResults).toMatchObject([
        { type: actionTypes.FETCH_USER_SUBSCRIPTIONS_REQUEST_VNEXT },
        {
          payload: expect.any(Object),
          type: actionTypes.FETCH_USER_SUBSCRIPTIONS_SUCCESS_VNEXT,
        },
      ]);
      expect(
        find(actionResults, {
          type: actionTypes.FETCH_USER_SUBSCRIPTIONS_SUCCESS_VNEXT,
        }),
      ).toMatchSnapshot('Fetch subscriptions success payload');
    });
  });
});
