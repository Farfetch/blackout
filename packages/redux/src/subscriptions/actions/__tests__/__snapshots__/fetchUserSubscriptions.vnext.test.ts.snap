// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscriptions redux actions fetchUserSubscriptionsVNext() action creator Should create the correct actions for when the get subscriptions is successful: Fetch subscriptions success payload 1`] = `
Object {
  "payload": Array [
    Object {
      "id": "8c2b5c3e3acb4bdd9c26ba46",
      "metadata": Object {
        "meta": "data",
      },
      "tenantId": 19934,
      "topics": Array [
        Object {
          "channels": Array [
            Object {
              "chanel": "None",
              "recipient": "1ca9c02be7e27f42bdfdca1afef2618003bbdc7d08fe2e9b54d2ac5af8b37127",
              "source": "My Account",
            },
            Object {
              "chanel": "Email",
              "recipient": "3c2b5c3e3acb4bdd9c26ba46",
              "source": "My Account",
            },
            Object {
              "chanel": "Sms",
              "recipient": "1234567890",
              "source": "My Account",
            },
          ],
          "name": "Any",
          "specification": Object {
            "type": "Any",
          },
        },
      ],
    },
  ],
  "type": "@farfetch/blackout-redux/FETCH_USER_SUBSCRIPTIONS_SUCCESS_VNEXT",
}
`;
